### PR TITLE
Fix / cleanup xrate computation

### PR DESCRIPTION
**Pull request by marcovc**
This PR:

* Finally fixes a long standing bug where the computed exchange rate would not be optimal for some corner cases.
* Refactors the xrate computation to be more in line with the documentation [here](https://github.com/gnosis/dex-open-solver/blob/master/doc/token_pair/token_pair.pdf).
* Optimizes the algorithm slightly (see chart below).
* Adds missing unit test for the algorithm that computes the xrate. Essentially tries to find a better exchange rate by a brute force algorithm.

image

### CODE REVIEW

File: src/core/config.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ class Config:
 
     # Convenience method to compute effective min tradable amount.
     @classproperty
```
Comment (twalth3r) [inline src/core/config.py:42]: Just curious, have never seen this decorator before - what does it do?

File: src/token_pair_solver/orderbook.py [hunk 2]
```diff
@@ -137,3 +137,12 @@ def aggregate_orders_prices(
 
 def count_orders_satisfying_xrate(b_orders, xrate, fee):
     return sum(xrate <= b_order.max_xrate * (1 - fee.value) for b_order in b_orders)
+
+
+def prune_unrealizable_orders(b_orders, s_orders, fee):
```
Comment (twalth3r) [inline src/token_pair_solver/orderbook.py:142]: A short comment might be helpful here.

File: tests/conftest.py [hunk 3]
```diff
@@ -15,6 +15,11 @@
     'debug', verbosity=hypothesis.Verbosity.verbose, max_examples=1000
 )
 
+# In the "paranoid" profile 10000 examples are generated.
```
Comment (twalth3r) [inline tests/conftest.py:18]: :rofl:

File: src/core/config.py (hunk 1 above)
Comment (marcovc) [reply src/core/config.py:42]: This one is a custom made (I copied the code from somewhere). It seems the "@property" decorator only applies to class "normal" members (those that have "self" as the first parameter). This one is a version of "@property" but for "@classmethod" members. This allows to do `Config.SOME_CONSTANT` even if that constant is not really a constant but instead a function that depends only on constants, that is, is a class method `SOME_CONSTANT(cls)`.

File: src/token_pair_solver/orderbook.py (hunk 2 above)
Comment (marcovc) [reply src/token_pair_solver/orderbook.py:142]: Thanks! The code still needs a lot of comments in general, and typing annotations. I intend to do that in a future PR. Add that one here df673b8.

**Comment by marcovc** [comment]
Thanks all for all the reviewing of something that very likely could me more well documented and structured.

I was trying to balance the effort/gain ratio, and since at this point we don't know if this "open solver" will be used by anyone else or not, I wasn't too worried about that part. I would like to add more comments and type annotations once we deal with the other urgent things. 

I'm sure there will be still issues to fix there, but after the unit tests added in this PR I'm confident that the impact of these potential bugs will be low.

File: src/core/config.py (hunk 1 above)
Comment (twalth3r) [reply src/core/config.py]: Thanks! :+1:

**Pull request merged**